### PR TITLE
 Support for MBTA's GTFS pathways 

### PR DESCRIPTION
**Pull request by lboyarsky**
Ticket: [add support for new pathways.txt specification](https://app.asana.com/0/810933294009540/1115551179980222)

Concerns:
- [Python code](https://github.com/mbta/gtfs_creator/blob/gtfs-creator2/gtfs_creator2/pathways/travel_time.py) linked to the ticket has support for calculating traversal times for slopes. It seems that we don't have this data for pathways as of now, so it's not implemented here.
- It seems that we currently don't have negative stair count anywhere despite the fact that [google spec](https://github.com/google/transit/blob/master/gtfs/spec/en/reference.md#pathwaystxt) says it should be negative for cases when a rider walks down the stairs. The walking speed is obviously different for cases when rider goes up and when they go down -- and this is supported in the python code -- but since we only have positive stair counts, we'll always assume that rider is going up
- I'm not fully sure how to test it, asked @jfabi in slack. All in all, pathways are displayed for the trip plans (but they used to be displayed before as well), but I'm not sure if they are displayed as expected: 

image
image

### CODE REVIEW

File: src/main/java/org/opentripplanner/model/Pathway.java [hunk 1]
```diff
@@ -57,16 +71,45 @@ public void setTraversalTime(int traversalTime) {
         this.traversalTime = traversalTime;
     }
 
-    public int getTraversalTime() {
-        return traversalTime;
+    public int calculateTraversalTime() {
```
Comment (lboyarsky) [inline src/main/java/org/opentripplanner/model/Pathway.java]: @jfabi would you be so kind to review business logic here and in `calculateWheelchairTraversalTime` below?

**Comment by paulswartz** [comment]
Can we use `signposted_as` (https://github.com/mbta/gtfs-documentation/blob/595f91b8353bd1d02c11acde50fc229584789724/reference/gtfs.md#pathwaystxt) to describe the pathway, instead of the generic `pathway`?

**Comment by lboyarsky** [comment]
Done:
Screen Shot 2019-04-16 at 10 47 26 AM

**Comment by paulswartz** [comment]
That's pretty awkward as text, but that seems like more an issue for website design :)

**Comment by jfabi** [comment]
@lboyarsky As far as testing goes, is it possible to write a short `pathways.txt` sample and compare, for a sample origin-destination paid, the output of the text to be displayed as a unit test? Is this what you are asking about, or something else?

**Regarding stairs**: That's my bad -- I'll have this fixed up in the next day or two.

**Regarding ramps/sloped pathways**: I hadn't implemented support for `max_slope` (a new field compared to how ramps were denoted previously), but it seems it actually would be much simpler than I anticipated, so I will likely add the field (populating it non-null for sloped ramps, null for all else) at the same time I fix `stair_count`.

**Regarding general display**: `signposted_as` sounds like the way to go. @paulswartz, should we have some content/design folks look at this here before we go live? As a separate task? For reference, Google (on mobile) uses the styling "Follow signs for [`signposted_as` for the final path in the sequence]".

@lboyarsky: I'm confused in the screenshots you have as the path is labeled "Left onto [`signposted_as`]." When I try the same plan in production, it tells me to go "Right onto pathway". I would expect to have to turn right to enter the station (though I also wonder how useful "Left" vs "Right" is anyway in this context).

Similarly, Davis is one of the stations for which we should have the expanded pathways with nodes (stairs, escalators, elevators). I only see one pathway listed entering the station, and no mention of any stairs, etc. Is this the intended behavior? When exiting Central in the screenshot provided, I also only see "Depart onto Transfer" -- should we be giving more precise pathways information here?

**Comment by paulswartz** [comment]
> **Regarding general display**: `signposted_as` sounds like the way to go. @paulswartz, should we have some content/design folks look at this here before we go live? As a separate task? For reference, Google (on mobile) uses the styling "Follow signs for [`signposted_as` for the final path in the sequence]".

Yes, we should file a ticket for website once we can put this up on dev-green for testing.

**Comment by lboyarsky** [comment]
So there were two issues:
1) The code responsible for linking street graph with pathways didn't take into account whether certain node is an entrance or not, so it was linking all the nodes directly to the streets. 
2) Some pathways, which have neither length nor time nor stair count, were considered to be unpassable. 

Here is how it looks now:
image

While this seems to be accurate in terms of path, I'm not sure whether this is something we want to present to the users as is. Any suggestions?

**Comment by paulswartz** [comment]
Can we use the description of the node itself, instead of/in addition to the pathway `signposted_as`?

**Comment by lboyarsky** [comment]
Yes, and this is how it looks like: 

image

I think it's less human-readable like that.

Also should we use elevators only for accessible trips may be?

**Comment by paulswartz** [comment]
You're right, that's definitely worse. Maybe @jfabi or Ingrid or Liana have ideas?

File: src/main/java/org/opentripplanner/model/Pathway.java [hunk 2]
```diff
@@ -7,9 +7,21 @@
 
     private static final int MISSING_VALUE = -999;
 
+    private static final int DEFAULT_TRAVERSAL_TIME = 5;
+
+    private static final int DEFAULT_WHEELCHAIR_TRAVERSAL_TIME = 10;
+
+    private static final float WALK_SPEED_FEET_SECONDS = 4.59f;
```
Comment (jfabi) [inline src/main/java/org/opentripplanner/model/Pathway.java]: The data that come from pathways.txt are in meters, not feet, so these numbers here have to be converted to be appropriate for meters.

**Comment by jfabi** [comment]
One thought, @lboyarsky @ingridpierre:

It would likely be better if we used a station like Beachmont or Orients Heights for the sample here. For the most recent stations being added to pathways.txt, I am filling out the `signposted_as` as it is intended (which is to say, what the signage at the station reads) -- this wasn't done for some of the earlier stations such as Davis, until we saw how Google Maps was using the field. So in reality, `signposted_as` might better read "Elevator" and "Red Line - All trains" instead of "Top of Elevator 817" and "Bottom of Elevator 817", respectively.

**Comment by jfabi** [comment]
A few concepts I had come up with (top left is what is currently being displayed in this PR):
FullSizeRender 15

The bottom concepts basically attempt to simplify the steps and providing more visual context based on the `pathway_mode` and the direction of travel for elevators/escalators/stairs -- changing the language for each mode of travel, or even just showing the vertical steps. @ingridpierre has also thought about making the pathways directions as a whole collapsable.

**Comment by ingridpierre** [comment]
can pathways be used to provide people alternate ways of navigating stations if they cant use the suggested escalators/stairs?

or are we just adding pathways for one specific way of navigating a station based on the most usual path?

**Comment by lboyarsky** [comment]
@ingridpierre trip planner will offer escalators/stairs only if you search for 'best route' (default) option; it's going to offer elevators for 'wheelchair accessible' search. Is this what you were asking about?

**Comment by jfabi** [comment]
@ingridpierre They can be used to provide alternate paths within a station. (See comment by @lboyarsky about displaying paths with elevators.) Another potential here could be done as a follow-up is to check if the path to be display has any Alerts associated with the escalators/elevators, and to choose an alternative route if that's the case (which could involve a different exit/entrance or station entirely).

File: src/main/java/org/opentripplanner/model/Pathway.java (hunk 2 above)
Comment (lboyarsky) [reply src/main/java/org/opentripplanner/model/Pathway.java]: Fixed, thanks.

**Comment by ingridpierre** [comment]
@lboyarsky @jfabi Ah, that makes sense. Can I explore these changes live on a staging site  somewhere? I'm curious about the examples Josh gave and a few others that are tricky to navigate like Kendall, Copley, Chinatown, etc.

I really like Josh's concepts above, but since design is required I want to make sure I have some examples well-documented before we make any recommendations.

Like he says above, incorporating alerts will also make this tricky, so it's useful to see live examples so I can think through those cases.

**Comment by lboyarsky** [comment]
> It would likely be better if we used a station like Beachmont or Orients Heights for the sample here.

**Beachmont** (same trip plans for accessible / inaccessible): 
image

**Orient Heights** (accessible):
image

**Comment by lboyarsky** [comment]
@ingridpierre deployed to https://green.dev.mbtace.com

**Comment by lboyarsky** [comment]
Doesn't seem that we're going to merge it anytime soon.

**Pull request closed without merging**